README.md
<!--
Original work Copyright (c) 2017 Dr. Frank Heimes (twitter.com/DrFGHde, www.facebook.com/dr.frank.heimes)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->

# WimLibScripts

WimLibScripts is a collection of PowerShell scripts to backup and restore entire Windows volumes.
It uses PowerShell to query all required settings and perform checks,
and the excellent open source library [WimLib](https://wimlib.net) to perform the actual backup and restore task.

The scripts can easily be configured to run fully interactively with file selection dialogs,
interactive but strictly text based, or fully automated with preset parameters.

## Motivation

Over the years, I had used *Acronis TrueImage*, *DriveImage XML*, *Norton GhostImage*, and the *Windows 7* built-in backup,
none of which fully satisfied me. So I stumbled over several articles that describe the features of [WimLib](https://wimlib.net),
which exploits the [WIM](https://de.wikipedia.org/wiki/Windows_Imaging_Format_Archive) file format
that is also used by Microsoft to bundle its software releases.

#### My initial experiments showed remarkable results:
 * the [WimLib](https://wimlib.net) library is rock solid and reliably creates and restores images.
 * it is ***very*** space efficient. An image consumes only about a third of the space used by the files on the volume.
 * it skips well-known space hoggers, like `pagefile.sys`, `swapfile.sys`, or `System Volume Information`, among others.
 * it adds additional images as differential changes to the WIM file.
   So the size of the WIM file only increases by about a third of the size of the modified files.
 * it is ***very*** CPU efficient. While compressing the file data in parallel, it used all of my four cores to 99%.
 * the generated WIM file can easily be browsed and extracted using the free [7-Zip](http://www.7-zip.org/) program.

Since the web pages I found merely describe the steps to manually issue a backup and how to restore an image,
I wrote these scripts to make this a regular, easy and less error prone task.

## Manual

For an explanation of the scripts, see the [MANUAL](MANUAL.md).

## License

Copyright (c) 2017 Dr. Frank Heimes  
See the [LICENSE](LICENSE.md) file for license rights and limitations (MIT).

## Support

Plaudits :smirk:, feature requests, and constructive comments are highly appreciated.
Since I'm using the scripts myself, I will continue to maintain them.

Feel free to contact me on [Twitter](https://twitter.com/DrFGHde) or [Facebook](https://www.facebook.com/dr.frank.heimes) :sunglasses:


Changes.md
<!--
Original work Copyright (c) 2017 Dr. Frank Heimes (twitter.com/DrFGHde, www.facebook.com/dr.frank.heimes)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->

# WimLibScripts CHANGES

## Version 1.1.17491.0

* Clean Manager and WIM file copy execute in parallel.
* Fixed mal-formed characters (hypen and space).

## Version 1.1.17347.0

* Support for volumes that do not support filesystem snapshots.

## Version 1.1.17242.0

* Made exclusion list explicit in backup.ini.

## Version 1.1.17201.0

* Fixed error message for WIM files on network shares.

## Version 1.1.17197.0

* Removed the `--solid` flag to make the resulting images mountable by DISM.
* Added scripts `MountImage.ps1` and `UnmountImage.ps1`

Many thanks to `synchronicity` for his helpful comments on the [WimLib Forum](https://wimlib.net/forums/viewtopic.php?f=1&t=310).

## Version 1.0.17114.1

* Initial version.


Restore.ps1
# Restore a volume, Version 1.1.17491.0
#
# Original work Copyright (c) 2017 Dr. Frank Heimes (twitter.com/DrFGHde, www.facebook.com/dr.frank.heimes)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

param([switch]$elevated)
if (!(New-Object Security.Principal.WindowsPrincipal $([Security.Principal.WindowsIdentity]::GetCurrent())).IsInRole([Security.Principal.WindowsBuiltinRole]::Administrator)) {
  if ($elevated) { Read-Host 'Failed to elevate' }
  else { Start-Process powershell.exe -Verb RunAs -ArgumentList ('-ExecutionPolicy Unrestricted -noprofile -nologo -file "{0}" -elevated' -f ($myinvocation.MyCommand.Definition)) }
  exit
}


# ---------------------- Configuration ----------------------

# Whether or not to run interactively.
#  W A R N I N G :  When setting this to $false, make sure all remaining parameters are correct!
#                   I'm really serious - You can annihilate your operating system with one click!
$Global:runInteractive = $true

# Whether or not to use file dialog boxes for more convenient file selection.
$Global:useFileDialogs = $true

# The target volume to restore. This can NOT be the live system.
$Global:volume = 'G:\'

# The source file to take the image from.
$Global:wimFile = 'M:\Backup\System.wim'

# The image index in the WIM file to restore. Use highest available if possible
$Global:imageIndex = 9999

# -----------------------------------------------------------


# Call wimlib-imagex with the specified arguments and throw on failure
function ImageX([string[]]$wimlibArgs)
{
	if (!(Test-Path '.\wimlib-imagex.exe'))
		{ Throw "wimlib-imagex.exe not found. Please download from http://wimlib.net" }

	''
	$errorCount = $Error.Count
	.\wimlib-imagex.exe $wimlibArgs
	if (($LastExitCode -ne 0) -or ($Error.Count -gt $errorCount))
		{ Throw "wimlib-imagex.exe failed" }
}

# Moves the window that executes this script to column 40 at the top of the screen
function Move-WindowToTop
{
	$Global:Win32SetWindowPos = Add-Type -memberDefinition '[DllImport("user32.dll")] public static extern bool SetWindowPos(IntPtr hWnd, IntPtr hWndInsertAfter, int X, int Y, int cx, int cy, int uFlags);' `
	-name 'Win32SetWindowPos' -namespace Win32Functions -passThru
	[void]$Global:Win32SetWindowPos::SetWindowPos(((Get-Process -Id $pid).MainWindowHandle), 0, 40, 0, 0, 0, 0x4255)
}

# Make the host window almost the height of the screen and place it at the top
function Configure-Host
{
	$MySize = $host.UI.RawUI.WindowSize
	$MySize.Height = $host.UI.RawUI.MaxPhysicalWindowSize.Height - 2
	$host.UI.RawUI.set_windowSize($MySize)
	Move-WindowToTop
}

# Explain what the script does and let user confirm execution
function Confirm-Execution
{
	"`n   R E S T O R E   V O L U M E`n"
	"This script will perform the following tasks:"
	"------------------------------------------------"
	" 1. Let you choose the name of a WIM file."
	" 2. Let you choose an image from the WIM file to restore."
	" 3. Let you choose which volume to restore."
	" 4. Verify the integrity of the WIM file."
	" 5. Quick format the volume as NTFS."
	" 6. Restore the selected image to the formatted volume."
	if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
}

# Let user select one of the available volumes to restore;
# The live volume, the CDRom, and the drive holding the WIM file is skipped.
function Select-Volume
{
	"`nSelect a volume to restore."
	"This must NEITHER be the live system ($env:SystemDrive), NOR the volume holding the WIM file!"
	# Name = Caption = DeviceID, Optionally interesting: VolumeSerialNumber
	if ($wimFile.StartsWith('\\')) { $qualifier = '0' } else { $qualifier = Split-Path -Qualifier $wimFile }
	Get-WmiObject Win32_logicaldisk | ?{ $_.Name -ne "${env:SystemDrive}" -and $_.Name -ne $qualifier -and $_.DriveType -eq 3 } | `
	Select Name,VolumeName,FileSystem,Description,@{Name="Size[GB]"; Expression={[Math]::Ceiling($_.Size / 1GB)}} | `
	Format-Table -AutoSize
	$answer = Read-Host "Enter the drive letter of the volume to restore ($($volume.TrimEnd(':\')))"
	if (![string]::IsNullOrEmpty($answer)) { $Global:volume = $answer }
	$Global:volume = $volume.ToUpper().TrimEnd(':\') + ':\'
}

# Let user select or enter a WIM file.
function Get-WIMFilename
{
	$initialFolder = Split-Path -Parent $wimFile
	"`nWIM files in ${initialFolder}:"
	"--------------------------------"
	if (Test-Path $initialFolder)
		{ Get-ChildItem -Path $initialFolder '*.wim' | Format-Table -Property FullName -AutoSize }
	$answer = Read-Host "Enter the full path of an existing backup file ($wimFile)"
	if (![string]::IsNullOrEmpty($answer)) { $Global:wimFile = $answer }
}

# Let user select a WIM file. Aborts script if user hits 'Abort'
function Get-WIMFilenameDlg
{
	[void][System.Reflection.Assembly]::LoadWithPartialName("System.windows.forms")
	$dialog = New-Object System.Windows.Forms.OpenFileDialog
	$dialog.initialDirectory = Split-Path -Parent $wimFile
	$dialog.Title = "Select an existing backup file"
	$initialFilename = [IO.Path]::GetFileNameWithoutExtension($wimFile)
	$dialog.FileName = $initialFilename
	$dialog.filter = "WIM (*.wim)| *.wim"
	[void]$dialog.ShowDialog()
	if ($initialFilename -eq $dialog.FileName) { Exit }
	$Global:wimFile = $dialog.FileName
}

# Determine highest image index in the WIM file
function Get-MaxImageIndex
{
	if (!(ImageX info, $wimFile | ?{ $_ -match 'Image Count:\s*(\d+)' }))
		{ Throw "Could not extract the number of images in file '$wimFile'" }

	$Global:imageIndex = ([int]$Matches[1], $imageIndex | Measure -Min).Minimum
}

# Let user select one of the images in the WIM file
function Select-ImageIndex
{
	ImageX info, $wimFile | ?{ $_ -notmatch`
	'(Architecture|Attributes|Boot Index|Build|Chunk Size|Compression|Default Language|Description|Display (Description|Name)|Edition ID|Flags|GUID|HAL|Hard Link Bytes|Installation Type|Languages|Modification Time|Part Number|Product (Suite|Type)|Service Pack Level|System Root|Version|WIMBoot compatible):' } | `
	%{ if ($_ -match '(Total Bytes:\s*)(\d+)') {
		if ([int64]$Matches[2] -ge 1GB) { $divider = 1GB ; $unit = " GB" } else { $divider = 1MB ; $unit = " MB" }
		$Matches[1] + [Math]::Ceiling($Matches[2] / $divider).ToString("N0") + $unit
		} else { $_ }}
	$answer = Read-Host "`nEnter 1-based 'Index' of the image to restore ($imageIndex)"
	if (![string]::IsNullOrEmpty($answer)) { $Global:imageIndex = [int]$answer }
}

# Test if the selected volume is the live volume, if yes, report this and exit script 
function Reject-LiveVolume
{
	if ($volume.TrimEnd(':\') -eq $env:SystemDrive.TrimEnd(':\'))
		{ Throw "You cannot restore to the live system $env:SystemDrive\" }
}

# Warn user if the target volume contains a page file. That may or may not be in use.
function Warn-AboutPageFile
{
	Get-WmiObject Win32_PageFileUsage | %{ $pageFilePath = $_.Name ; [int]$pageFileUse = $_.CurrentUsage }
	$pageFile = $volume.TrimEnd(':\') + ":\pagefile.sys"
	if ((Test-Path $pageFile) -and ($pageFilePath -eq $pageFile) -and ($pageFileUse -gt 0)) {
		"`nWarning: The currently active paging file '$pageFile' may render the volume '$volume' read-only."
		"In the advanced system settings, make sure that '$pageFile' is not ACTIVELY being used."
		"Verify again that '$volume' is the correct volume!"
		if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
	}
}

# Warn user if the target volume is too small to hold the entire image.
function Warn-AboutSmallVolume
{
	[int64]$volumeSize = Get-WmiObject Win32_logicaldisk | ?{ $_.Name -eq $volume.TrimEnd('\') } | %{ $_.Size }
	[int64]$imageSize = 0
	if (ImageX info, $wimFile, $imageIndex | ?{ $_ -match 'Total Bytes:\s*(\d+)' })
		{ $imageSize = $Matches[1] }

	if ($imageSize -gt $volumeSize) {
		"`nWarning: The volume '$volume' might be too small to hold the entire image."
		"The volume size is just $([Math]::Round($volumeSize / 1GB,0).ToString('N0')) GB, but the image size may be up to $([Math]::Ceiling($imageSize / 1GB).ToString('N0')) GB."
		"You will be able to extract some of the image files but the"
		"volume may not be complete and it might be corrupted."
		if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
	}
}

# Let user confirm all selected parameters before commencing operation. Abort script if user does not confirm.
function Confirm-Parameters
{
	"`nVolume '$volume' will be formatted. Then image '$imageIndex' from the"
	"WIM file '$wimFile' will be applied to the volume."
	"`n W A R N I N G"
	"   All data currently on volume '$volume' will be deleted permanently!"
	if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
	"`nNote: You can still abort the script (at your own risk!) as long as the volume formatting has not begun."
}

# Verifies the integrity of the WIM file
function Verify-WIMFile
{
	ImageX verify, $wimFile
}

# Format the target volume. Throw on failure.
function Format-TargetVolume
{
	$volumeName = Get-WmiObject Win32_logicaldisk | ?{ $_.Name -eq $volume.TrimEnd('\') } | %{ $_.VolumeName }
	if ([string]::IsNullOrEmpty($volumeName)) { $volumeName = 'Restored' }
	"`nFormating volume '$volume' ($volumeName) ..."
	$errorCount = $Error.Count
	Format-Volume -DriveLetter $volume.TrimEnd(':\') -FileSystem 'NTFS' -NewFileSystemLabel $volumeName -Force
	if (($LastExitCode -ne 0) -or ($Error.Count -gt $errorCount))
		{ Throw "Format-Volume failed" }

	# Without this delay, the summary of Format-Volume appear later in the output, mixed with other messages.
	Start-Sleep 1
}

# Format the target partition before restoring the contents
function Restore-Volume
{
	"`nRestoring image $imageIndex from '$wimFile' to partition '$volume' ..."
	ImageX apply, $wimFile, $imageIndex, $volume
}

# Compute and report the execution time w.r.t. the provided start time
function Report-ExecutionTime([DateTime]$startTime)
{
	"`nRestore ran for " + ((Get-Date) - $startTime).ToString("h\:mm\:ss")
}

# Explain script, collect imputs from user and perform restore operation
function Run-Restore
{
	Try {
		if ($runInteractive) {
			Configure-Host
			Confirm-Execution
			if ($useFileDialogs) { Get-WIMFilenameDlg } else { Get-WIMFilename }
		}
		Get-MaxImageIndex
		if ($runInteractive) {
			Select-ImageIndex
			Select-Volume
		}
		Reject-LiveVolume
		if ($runInteractive) {
			Warn-AboutPageFile
			Warn-AboutSmallVolume
			Confirm-Parameters
		}
		$startTime = Get-Date
		Verify-WIMFile
		Format-TargetVolume
		Restore-Volume
		Report-ExecutionTime $startTime
		if ($runInteractive)
			{ Read-Host "`nDone. Press Enter to finish" }
	}
	Catch {
		Write-Error $_
		if ($runInteractive)
			{ Read-Host "`nPress Enter to abort" }
	}
}

Set-Location ($MyInvocation.InvocationName | Split-Path -Parent)
Run-Restore


Backup.ps1
# Backup a volume, Version 1.1.17491.0
#
# Original work Copyright (c) 2017 Dr. Frank Heimes (twitter.com/DrFGHde, www.facebook.com/dr.frank.heimes)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

param([switch]$elevated)
if (!(New-Object Security.Principal.WindowsPrincipal $([Security.Principal.WindowsIdentity]::GetCurrent())).IsInRole([Security.Principal.WindowsBuiltinRole]::Administrator)) {
  if ($elevated) { Read-Host 'Failed to elevate' }
  else { Start-Process powershell.exe -Verb RunAs -ArgumentList ('-ExecutionPolicy Unrestricted -noprofile -nologo -file "{0}" -elevated' -f ($myinvocation.MyCommand.Definition)) }
  exit
}


# ---------------------- Configuration ----------------------

# Whether or not to run interactively.
#  W A R N I N G :  When setting this to $false, make sure all remaining parameters are correct!
$Global:runInteractive = $true

# Whether or not to use file dialog boxes for more convenient file selection.
$Global:useFileDialogs = $true

# Whether or not to run the Windows Clean Manager before creating backup
$Global:runCleanMgr = $true

# Whether or not to copy an existing WIM file before modifying it
$Global:copyWIMFile = $true

# The source volume to backup. This may be the live system, too.
$Global:volume = 'C:\'

# The target file to add the image to. Preferably on a mounted external USB drive.
$Global:wimFile = 'M:\Backup\System.wim'

# The name for the new image in the backup file. Defaults to the date and time.
$Global:imageName = (Get-Date).DateTime

# -----------------------------------------------------------


# Call wimlib-imagex with the specified arguments and throw on failure
function ImageX([string[]]$wimlibArgs)
{
	if (!(Test-Path '.\wimlib-imagex.exe'))
		{ Throw "wimlib-imagex.exe not found. Please download from http://wimlib.net" }

	''
	$errorCount = $Error.Count
	.\wimlib-imagex.exe $wimlibArgs
	if ($LastExitCode -eq 89)
		{ "`r`nThe selected volume does not support filesystem snapshot - retrying without ..." }
	if (($LastExitCode -ne 0) -or ($Error.Count -gt $errorCount))
		{ Throw "wimlib-imagex.exe failed" }
}

# Moves the window that executes this script to column 40 at the top of the screen
function Move-WindowToTop
{
	$Global:Win32SetWindowPos = Add-Type -memberDefinition '[DllImport("user32.dll")] public static extern bool SetWindowPos(IntPtr hWnd, IntPtr hWndInsertAfter, int X, int Y, int cx, int cy, int uFlags);' `
	-name 'Win32SetWindowPos' -namespace Win32Functions -passThru
	[void]$Global:Win32SetWindowPos::SetWindowPos(((Get-Process -Id $pid).MainWindowHandle), 0, 40, 0, 0, 0, 0x4255)
}

# Make the host window almost the height of the screen and place it at the top
function Configure-Host
{
	$MySize = $host.UI.RawUI.WindowSize
	$MySize.Height = $host.UI.RawUI.MaxPhysicalWindowSize.Height - 2
	$host.UI.RawUI.set_windowSize($MySize)
	Move-WindowToTop
}

# Explain what the script does and let user confirm execution
function Confirm-Execution
{
	"`n   B A C K U P   V O L U M E`n"
	"This script will perform the following tasks:"
	"------------------------------------------------"
	" 1. Query the volume to backup."
	" 2. Query the name of a WIM backup file."
	" 3. Query a name for this backup."
	" 4. Optionally make a copy of the WIM file if it exists."
	" 5. Optionally call the Clean Manager to remove obsolete data."
	"     You should run it once like this to configure what to delete:"
	"       'cleanmgr /sageset:1'"
	" 6. Create or update the backup file."
	" 7. Verify the integrity of the backup file."
	" 8. Create a log file with a list of files in the new image."
	if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
}

# Let user select one of the available volumes to backup; CDRom volumes are skipped.
function Select-Volume
{
	# Name = Caption = DeviceID, Optionally interesting: VolumeSerialNumber
	Get-WmiObject Win32_logicaldisk | ?{ $_.DriveType -eq 3 } | Select Name,VolumeName,`
	@{Name="Total[GB]"; Expression={[Math]::Ceiling($_.Size / 1GB)}},`
	@{Name="Used[GB]"; Expression={[Math]::Ceiling(($_.Size - $_.FreeSpace) / 1GB)}},FileSystem,Description | `
	Format-Table -AutoSize
	"Enter the drive letter of the volume to backup."
	$answer = Read-Host "This can be the live system, too ($($volume.TrimEnd(':\')))"
	if (![string]::IsNullOrEmpty($answer)) { $Global:volume = $answer }
	$Global:volume = $volume.ToUpper().TrimEnd(':\') + ':\'
}

# Let user select or enter a WIM file.
function Get-WIMFilename
{
	$initialFolder = Split-Path -Parent $wimFile
	"`nWIM files in ${initialFolder}:"
	"--------------------------------"
	if (Test-Path $initialFolder)
		{ Get-ChildItem -Path $initialFolder '*.wim' | Format-Table -Property FullName -AutoSize }
	$answer = Read-Host "Enter the full path of an existing or new backup file ($wimFile)"
	if (![string]::IsNullOrEmpty($answer)) { $Global:wimFile = $answer }
}

# Let user select or enter a WIM file. Aborts script if user hits 'Abort'
function Get-WIMFilenameDlg
{
	[void][System.Reflection.Assembly]::LoadWithPartialName("System.windows.forms")
	$dialog = New-Object System.Windows.Forms.SaveFileDialog
	$dialog.OverwritePrompt = $false
	$dialog.initialDirectory = Split-Path -Parent $wimFile
	$dialog.Title = "Select an existing backup file or select a folder and enter a new name"
	$initialFilename = [IO.Path]::GetFileNameWithoutExtension($wimFile)
	$dialog.FileName = $initialFilename
	$dialog.filter = "WIM (*.wim)| *.wim"
	[void]$dialog.ShowDialog()
	if ($initialFilename -eq $dialog.FileName) { Exit }
	$Global:wimFile = $dialog.FileName
}

# Test if the WIM file is on the volume that is to be backed up, if yes, report this and exit script 
function Reject-VolumeConflict
{
	if (!$wimFile.StartsWith('\\') -and $volume -eq "$(Split-Path -Qualifier $wimFile)\") {
		Throw "The backup of volume '$volume' cannot be written to the WIM file '$wimFile' on that same volume."
	}
}

# Let user override the name for this image
function Get-ImageName
{
	"`nAll images in a WIM file must have unique names,"
	"which should not contain special characters."
	$answer = Read-Host "`nEnter a name for this backup image ($imageName)"
	if (!([string]::IsNullOrEmpty($answer))) { $Global:imageName = $answer }
}

# Query whether or not to run Windows Clean Manager
function Query-CleanMgr
{
	$Global:runCleanMgr = (Read-Host "`nRun Clean Manager (Y/N)?") -eq 'y'
}

# Run Windows Clean Manager if configured
function Start-CleanMgr
{
	if ($runCleanMgr) {
		$cleanJob = Start-Job { cleanmgr /sagerun:1 | Out-Null } # | Out-Null waits for clean manager to finish
	}
}

# Wait for Windows Clean Manager if started
function Wait-CleanMgr
{
	if ($runCleanMgr -and $cleanJob -ne $null) {
		Wait-Job $cleanJob
	}
}

# Test if the specified file exists, if yes, report this and exit script 
function Reject-ExistingBackup([string]$wimFileBackup) 
{
	if (Test-Path $wimFileBackup) {
		Throw "File '$wimFileBackup' already exists.`n" + 
		"Rename or delete it first and try again."
	}
}

# Query whether or not to create a backup of an existing WIM file
function Query-WIMFileBackup([string]$wimFileBackup)
{
	if ($copyWIMFile)
		{ $Global:copyWIMFile = (Read-Host "`nCopy '$wimFile' as '$wimFileBackup' (Y/N)?") -eq 'y' }
}

# Creates a copy of the WIM file after user confirmation
function Create-WIMFileBackup([string]$wimFileBackup)
{
	if (!$copyWIMFile)
		{ return }

	Reject-ExistingBackup $wimFileBackup
	Import-Module BitsTransfer
	Start-BitsTransfer -Source $wimFile -Destination $wimFileBackup `
		-Description "Copying '$wimFile' as '$wimFileBackup'" -DisplayName "Copying backup file"
}

# Let user confirm all selected parameters before commencing operation. Abort script if user does not confirm.
function Confirm-Parameters([string]$wimFileBackup)
{
	''
	if ($runCleanMgr) { "The Windows Clean Manager will be called." }
	if ($copyWIMFile) { "The existing WIM file '$wimFile' will be copied`nas '$wimFileBackup'." }
	"A backup of volume '$volume' with the name '$imageName'"
	"will be added to the WIM file '$wimFile'"
	if ((Read-Host "`nContinue (Y/N)?") -ne 'y') { Exit }
}

# $true if a new WIM file has been created, $false if an existing one was updated
$Global:newWIMFile = $false

# Creates a named image of the specified volume in the specified WIM file
function Create-Image
{
	if (Test-Path $wimFile) {
		"`nUpdating backup file '$wimFile' ..."
		$command = 'append'
		$Global:newWIMFile = $false
	} else {
		"`nCreating new backup file '$wimFile' ..."
		$command = 'capture'
		$Global:newWIMFile = $true
	}

	# Optional: Add option  '--solid'  to create a solid archive.
	#           As that produces somewhat smaller files, these cannot be mounted (limitation of DISM).
    [Collections.ArrayList]$wimlibArgs = $command, $volume, $wimFile, $imageName, '--boot', '--check', '--snapshot', '--config=Backup.ini'

    Try
    {
	    $ErrorActionPreference = 'Stop'
        # Use the snapshot function. Active file systems require this.
	    ImageX $wimlibArgs
    }
    Catch
    {
        # Logic file systems don't support --snapshot, so try again without that option and let exceptions pass.
        $wimlibArgs.Remove('--snapshot')
    	ImageX $wimlibArgs
    }
}

# Verifies the integrity of the WIM file
function Verify-WIMFile
{
	ImageX verify, $wimFile
}

# Get the index of the latest image from the backup file. Abort script on failure
function Get-ImageIndex
{
	if (ImageX info, $wimFile | ?{ $_ -match 'Image Count:\s*(\d+)' })
		{ return $Matches[1] }

	Throw "Could not extract the number of images in file '$wimFile'`n" +
	"Please check the contents of the file yourself."
}

# Create a file list (i.e. table of contents) of the WIM file into a text file
function Create-FileList
{
	$listFile = Join-Path (Split-Path -Parent $wimFile) $([IO.Path]::GetFileNameWithoutExtension($wimFile) + "$imageIndex.txt")
	"`nListing files of image $imageIndex in '$wimFile' to '$listFile' ..."
	ImageX dir, $wimFile, $imageIndex > $listFile
	"Listed files of image $imageIndex into '$(Resolve-Path $listFile)'"
	"You can add a description of the volume changes at the beginning of that file using a text editor."
}

# Compute and report the execution time w.r.t. the provided start time
function Report-ExecutionTime([DateTime]$startTime)
{
	"`nBackup ran for " + ((Get-Date) - $startTime).ToString("h\:mm\:ss")
}

# Inform user that he can delete 
function Final-Message([string]$wimFileBackup)
{
	if ($newWIMFile) {
		"`nSince this is the first backup of volume '$volume', you are highly encouraged"
		"to restore it right away to make sure that everything works smoothly."
		"That includes booting the Windows on a USB drive or an installation CD."
	}
	
	if (Test-Path $wimFileBackup) {
		"`nYou may now move or delete the WIM backup file '$wimFileBackup'"
		"if the updated WIM file '$wimFile' was verified successfully."
	}
}

# Explain script, collect imputs from user and perform backup operation
function Run-Backup
{
	Try {
		if ($runInteractive) {
			Configure-Host
			Confirm-Execution
			Select-Volume
			if ($useFileDialogs) { Get-WIMFilenameDlg } else { Get-WIMFilename }
			Reject-VolumeConflict
			Get-ImageName
			Query-CleanMgr
		}
		$wimFileName = $([IO.Path]::GetFileNameWithoutExtension($wimFile))
		$wimFileBackup = Join-Path (Split-Path $wimFile -Parent) "$wimFileName-$((Get-Date).ToShortDateString()).wim"
		$Global:copyWIMFile = (Test-Path $wimFile)
		if ($runInteractive) {
			Query-WIMFileBackup $wimFileBackup
			Confirm-Parameters $wimFileBackup
		}
		Start-CleanMgr
		Create-WIMFileBackup $wimFileBackup
		Wait-CleanMgr
		$startTime = Get-Date
		Create-Image
		Verify-WIMFile
		$Global:imageIndex = Get-ImageIndex
		Create-FileList
		Report-ExecutionTime $startTime
		Final-Message $wimFileBackup
		if ($runInteractive)
			{ Read-Host "`nDone. Press Enter to finish" }
	}
	Catch {
		Write-Error $_
		if ($runInteractive)
			{ Read-Host "`nPress Enter to abort" }
	}
}

Set-Location ($MyInvocation.InvocationName | Split-Path -Parent)
Run-Backup
